### PR TITLE
Avoid printing comments in SMT output, unless logging queries

### PR DESCRIPTION
**Pull request by nikswamy**
I noticed that printing comments (aka captions) in the SMT output blows up the output size enormously. E.g., in TLSConstants, with the new Parser.CipherSuites, the size of the SMT output is 79MB with comments included but only 9MB without. 

So, if log_queries is off, we don't print comments any more. 

And even with log_queries on, a new option `--keep_query_captions false` can be used to turn off printing of comments.

Further, I added support for printing the size of a module's encoding to SMT in the SMT output (see the way the new `Module` construct is printed in SMTEncoding.Term.fs

This also means that with log_queries off, we do not have to print to SMT twice (once without comments for taking the hash, and once with comments). That's the part that I'd appreciate a second set of eyes on ... i.e., that we hash SMT queries without relying on local details (like filenames etc.).

### CODE REVIEW

File: src/smtencoding/FStar.SMTEncoding.Z3.fs [hunk 1]
```diff
@@ -546,26 +546,22 @@ let mk_input theory =
...
-            | DeclFun (n, a, s, _) -> DeclFun (n, a, s, None)
-            | DefineFun (n, a, s, b, _) -> DefineFun (n, a, s, b, None)
-            | d -> d
+            let hs =
+              if Options.log_queries()
+              && Options.keep_query_captions ()
+              then prefix
+                   |> List.map (declToSmt_no_caps options)
+                   |> String.concat "\n"
+              else ps
```
Comment (aseemr) [inline src/smtencoding/FStar.SMTEncoding.Z3.fs]: Looks fine to me. A minor nit: we could move `Options.log_queries () && Options.keep_query_options ()` to a single function `print_query_captions` or something, and use it here as well as in the definition of `declToSMT`.

Comment (aseemr) [reply src/smtencoding/FStar.SMTEncoding.Z3.fs]: One way to check correctness here is that at this point, `ps = prefix  |> List.map (declToSmt options) |> String.concat "\n"`. And when `Options.log_queries () && Options.keep_query_options () = false`, `declToSmt = declToSmt_no_caps`.

File: src/smtencoding/FStar.SMTEncoding.Term.fs [hunk 2]
```diff
@@ -717,21 +717,21 @@ let termToSmt
...
 
-let rec declToSmt' print_ranges z3options decl =
+let rec declToSmt' print_captions z3options decl =
   match decl with
   | DefPrelude ->
     mkPrelude z3options
   | Module (s, decls) ->
-    let res = List.map (declToSmt' print_ranges z3options) decls |> String.concat "\n" in
+    let res = List.map (declToSmt' print_captions z3options) decls |> String.concat "\n" in
     if Options.log_queries ()
```
Comment (aseemr) [inline src/smtencoding/FStar.SMTEncoding.Term.fs]: Should this be Options.log_queries () && Options.keep_query_captions ()? Else the query hash will depend on the string below (if log_queries is on and keep_query_options is off).

**Comment by mtzguido** [comment]
Looks good to me, and I agree with Aseem on the tiny refactoring and the point about hashes (I confirmed they change with `--keep_query_captions`).

**Comment by nikswamy** [comment]
Thanks for the review! Will make these changes now.

File: src/smtencoding/FStar.SMTEncoding.Z3.fs (hunk 1 above)
Comment (nikswamy) [reply src/smtencoding/FStar.SMTEncoding.Z3.fs]: I didn't understand your second remark, @aseemr. Can you say more, please?

Comment (aseemr) [reply src/smtencoding/FStar.SMTEncoding.Z3.fs]: Oh nothing much, was just trying to convince myself that in the `else` branch, `ps` is already same as `prefix |> List.map (declToSmt_no_caps options) |> String.concat "\n"`.

**Pull request closed without merging**